### ISSUE TITLE
references/sources

### ISSUE THREAD

**Issue by jonsterling** [jonsterling/algebraic-foundations-of-proof-refinement#18]
I need to add references for each of the following topics:

- [ ] something for multi-sorted lawvere theories (maybe Taylor or Jacobs?)
- [ ] lax natural transformations, prestacks
- [ ] LCF: cite the original LCF stuff, but also Cambridge LCF probably too
- [ ] ...

**Comment by freebroccolo** [comment]
I haven't seen any references to multi-sorted or colored Lawvere theories from Jacobs or Taylor before. I think you'd have better luck looking for discussion of that in material from Leinster or Shulman discussing generalized multicategories or in material from Yau or Hackney discussing colored props. Actually there might be something in one of Martin Hyland's papers talking about Lawvere theories too.

It seems to me that it's more common to either go for colored operads instead or just skip straight to something potentially more general à la Leinster/Shulman.

**Comment by jonsterling** [comment]
There is something in the nlab somewhere (I'll link to it when I get back to my office); I think the basic idea is probably simple enough that I don't need to dwell on it---another perspective is that I am working directly with a classifying category to avoid fixing too much in advance.

**Comment by freebroccolo** [comment]
Ah, I see it. Linking to that might be enough. It's the only real explanation I've seen of the idea anywhere as far as I can remember. These kind of things don't really come up that much in the literature. Even discussion of colored operads is fairly rare.

**Comment by jonsterling** [comment]
@freebroccolo Yeah, that makes sense... The thing I was thinking of was here: https://ncatlab.org/nlab/show/algebraic+theory#multisorted_algebraic_theories.

**Comment by freebroccolo** [comment]
Okay, that's different than the one I was just looking at [here](https://ncatlab.org/toddtrimble/published/multisorted+Lawvere+theories).

**Comment by jonsterling** [comment]
@freebroccolo Ah, I didn't realize they had a whole page about it. That is probably a better source...

**Comment by freebroccolo** [comment]
Another take on it is that it might be better (or at least more accessible) to frame a discussion of Lawvere theories in terms of cartesian operads (as Trimble describes on that pages and [linked pages](https://ncatlab.org/toddtrimble/published/Towards+a+doctrine+of+operads#cartesian_operads_are_equivalent_to_lawvere_theories)). But I'm kind of biased toward the language of operads for various reasons so ymmv. On that page Trimble also seems to suggest that cartesian operads are basically the same thing as clones.

**Comment by jonsterling** [comment]
@freebroccolo I like the idea of operads a lot—the presentation on that nlab page is really nice and clear...

On the other hand, I'm mostly using this as a technical device to simplify the presentation (which will probably have to be read by PL and type theory people), and make it easily instantiatable under different notions of syntax—for instance, whilst my whole apparatus appears to be first-order (in the sense that the language of evidence doesn't have binding), by making `T` the classifying category for a second-order theory (as defined by Fiore & Mahmoud), the thing can trivially be used with a second-order notion of evidence.

So, I'm hoping to avoid centering the specific technical/categorical implementation of this idea and just assume a category from the start that has enough structure to do what I want to do. Maybe there is a way to phrase what I say in the introduction that avoids some of these questions...

 What do you think?

**Comment by freebroccolo** [comment]
You'll have to refresh my memory a bit on the terminology. Is the classifying category the same thing as the free cartesian category over the arity graph for the language? If this is indeed the case, then there is a direct analogue in terms of a "classifying operad" which I think may be what you were looking for in the Lawvere theory thing. If that's the right idea, I don't see why there would be any technical reason why that couldn't be generalized to the second order case.

**Comment by jonsterling** [comment]
the classifying category is basically the category of contexts and substitutions. Does that help?

**Comment by freebroccolo** [comment]
Ah, so you need products in the codomain of some cells too then? That sounds more like a colored PROP (which is also nice to work with). Is that what you were referring to after you had the discussion with Awodey?

**Comment by jonsterling** [comment]
yeah... having the uniformity of products on both sides simplifies a lot of the bureaucratic aspects of my development. additionally, this category is very nice to do presheaves/prestacks over, which is our main maneuver (to induce substitution of evidence in judgments).

steve mentioned PROPs as a possibly-non-cartesian (e.g. monoidal) version of the classifying cat.

**Comment by freebroccolo** [comment]
Okay, I think I understand better now. So one thing you could do is just *start* with colored PROPs and add the morphisms to make it cartesian. I believe this gives you colored Lawvere theories directly in a way that's probably close to what you're already thinking about. But I don't know if having that is particularly useful or not.

**Comment by freebroccolo** [comment]
I mean, that's assuming you are wanting to use Lawvere theories in some sense. I need to go reread what you've got now.

**Comment by freebroccolo** [comment]
Oh I see you're already using them :grin:

**Comment by freebroccolo** [comment]
The use of Lawvere theories seems good.

The reason why I was suggesting it could be useful to start with PROPs is because the presentation for PROPs is so simple that you could just define the entire concept (multi-sorted Lawvere theories) in the paper and not worry about references (since there doesn't seem to be anything published).

A colored PROP is just a category where the domain/codomain are lists of objects and with *two* kinds of composition: the normal categorical one (vertical composition) and the whiskering/functoriality for the product (horizontal composition).

So you could start with that and just add another operation "pi" that would give you all cartesian projections. I think that's a pretty clean way to present the idea. The coherence laws are a little bit more complex but probably reasonable enough to fit into an appendix.

**Comment by jonsterling** [comment]
@freebroccolo Ah, interesting, I think I see what you're getting at now... The definition of PROPs on the nlab is really nice and clean! It seems closer to the sort of thing that I like to use than an operad...

It might even be not too difficult to simply define the notion of a "colored Lawvere theory" from scratch, following the definition of PROP (but not using), rather just requiring that our thing be a strict cartesian category whose monoid of objects is given by the set of sorts. On the other hand, it might suffice to just assume that the reader knows what is meant by "multi-sorted Lawvere theory", considering the fact that the entire construction appears to be doable in less than a paragraph of text.

(Another option would be to avoid mentioning any of this apparatus and simply fix a signature at the beginning, and say that we will work with its classifying category. Not sure whether that is better or worse!)

**Comment by jonsterling** [comment]
BTW, this paper very briefly defines colored lawvere theories in passing on page 9: https://arxiv.org/pdf/1601.00301.pdf

I suppose it might also be interesting to ask whether my apparatus actually needs the cartesian structure, or whether something substructural like a PROP would suffice.

**Comment by freebroccolo** [comment]
Interesting find. I think I've seen that paper from somewhere before but can't remember where.

As for whether or not you need the cartesian structure, I wouldn't be surprised if you didn't but I'd imagine that having it could make certain parts easier.

By the way, here is a sketch of what I was describing with the PROP approach:

```agda
module Perm where
  infix 6 ⇑_

  data Coselect {I} A : List I → List I → Set where
    ●
      : ∀ {Γ}
      → Coselect A (A ∷ Γ) Γ
    ⇑_
      : ∀ {B Γ Δ}
      → Coselect A Γ Δ
      → Coselect A (B ∷ Γ) (B ∷ Δ)

  Loc : ∀ {I} (Γ : Seq I) (A : I) (Δ : Seq I) → Set
  Loc Γ A Δ = Coselect A Γ Δ
  {-# DISPLAY Coselect A Γ Δ = Loc Γ A Δ #-}

  module 𝔖 where
    infixr 5 _∷_

    data 𝔖 {I} : List I → List I → Set where
      []
        : 𝔖 [] []
      _∷_
        : ∀ {A Γ Δ Θ}
        → Loc Γ A Δ
        → 𝔖 Δ Θ
        → 𝔖 Γ (A ∷ Θ)

    idn : ∀ {I Γ} → 𝔖 {I} Γ Γ
    idn {Γ = []} = []
    idn {Γ = A ∷ Γ} = ● ∷ idn

  open 𝔖 public
    hiding (module 𝔖)
    hiding (idn)
open Perm public

record Lawvere : Set where
  no-eta-equality
  field
    obj : Set
    hom : List obj → List obj → Set
  field
    idn : ∀ {Γ} → hom Γ Γ 
    seq : ∀ {Γ Δ Θ} → hom Γ Δ → hom Δ Θ → hom Γ Θ  
    wsk
      : ∀ {Γ₀ Γ₁ Δ₀ Δ₁}
      → (γ : hom Γ₀ Γ₁)
      → (δ : hom Δ₀ Δ₁)
      → hom (Γ₀ ++ Δ₀) (Γ₁ ++ Δ₁)
    swp -- two-sided permutation
      : ∀ {Γ Δ Θ K}
      → 𝔖 Γ Δ -- permute before
      → hom Δ Θ
      → 𝔖 Θ K -- permute after
      → hom Γ K
    nth
      : ∀ {Γ}
      → (i : Fin (length Γ))
      → hom Γ (lookup i Γ ∷ [])
```

**Comment by freebroccolo** [comment]
You could also simplify the presentation by using definition of permutations which can be given with the ◇ modality and bijections.

**Comment by jonsterling** [comment]
Aha, I think I *do* need the theory to be cartesian—weakening is essential in some of the proofs (and probably definitions).

**Comment by jonsterling** [comment]
- [ ] Ahh, a good thing to cite for lax natural transformations is Leinster's book, [Higher Operads, Higher Categories](https://arxiv.org/pdf/math/0305049v1.pdf)
